libcalico-go/lib/backend/k8s/resources/ipam_affinity.go

```diff
@@ -264,8 +382,8 @@ func (c *blockAffinityClient) List(ctx context.Context, list model.ListInterface
 		}
 		if host == "" || v1kvp.Key.(model.BlockAffinityKey).Host == host {
 			cidr := v1kvp.Key.(model.BlockAffinityKey).CIDR
-			cidr2 := &cidr
-			if requestedIPVersion == 0 || requestedIPVersion == cidr2.Version() {
+			cidrPtr := &cidr
+			if (requestedIPVersion == 0 || requestedIPVersion == cidrPtr.Version()) && !v1kvp.Value.(*model.BlockAffinity).Deleted {
```

Comment:
I think this might be problematic. Think of this situation:

- Perform a List()
- Start a Watch()
- Receive a deleted event for a block that wasn't present in the List(). 

Might just be better to show "Deleted" affinities in the list.

Reply:
The whole `Get`, `List`, `Watch` flow is a little strange. If we make `List` and `Watch` consistent, then `Get` and `List` are not consistent (you can try to `Get` an affinity you just deleted and not see it, but then might notice it showing up in a `List`). My take is that `Watch` is always going to look a little funny because unless you understand what is happening in the backend and why, every single time you delete an affinity, you'll see 2 events: an update and a delete. I feel like it is probably more useful for `List` and `Get` to be more consistent since I picture them being more run by users from the command line vs `Watch` which I envision will be some process or logging that is monitoring affinity activity.

Reply:
I think for the v3 API we have the opportunity to make all three consistent - it's only the v1 client that needs to delete-on-get-if-deleted. 

Sine v3 is read-only, we should just make watch / get / list all consistent with each other. Probably by not filtering out "Deleted=true" at all.

Reply:
I think the one downside of making the v3 client behavior differ from the v1 client behavior is that we will not be able to eventually move the IPAM logic to use the v3 client because of the behavior mismatch. Not a big deal but it does mean we may have to change API behavior on users if we decide to go that route later.

Reply:
Yeah, that is a downside, but I don't see how we get around that. We probably just need to keep the two clients separate and deal with the fact that that's the way it is.

Otherwise, we're intentionally baking in a defective API experience for affinities into the v3 API in order to satisfy cleaner code